src/render/glsl.rs

```diff
@@ -12,28 +12,29 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::collections::HashMap;
+use std::{borrow::Cow, collections::HashMap};
 
 #[derive(Default)]
 pub struct Registry {
-    shaders: HashMap<String, String>,
+    shaders: HashMap<&'static str, Cow<'static, str>>,
     shader_version: String,
 }
 
 impl Registry {
     pub fn new(shader_version: &str) -> Registry {
         Registry {
             shaders: Default::default(),
-            shader_version: shader_version.to_string(),
+            shader_version: shader_version.into(),
         }
     }
 
-    pub fn register(&mut self, name: &str, source: &str) {
+    pub fn register(&mut self, name: &'static str, source: impl Into<Cow<'static, str>>) {
+        let source = source.into();
         if self.shaders.contains_key(name) {
             panic!("shader {} is already defined", name);
         }
         self.shaders
-            .insert(name.to_owned(), source.trim().to_owned());
```

Comment:
Are you sure that removing this trim call isn't causing a change in behavior?

Reply:
Any trailing whitespace should be insignificant in GLSL